lib/wibox/widget/calendar.lua

```diff
@@ -0,0 +1,405 @@
+        spacing     = props.spacing,
+        forced_num_rows = num_rows,
+        forced_num_cols = num_columns,
+    }
+
+    local start_row    = 3
+    local start_column = num_columns - 6
+    local week_start   = props.start_sunday and 1 or 2
+    local d            = os.date("*t", os.time{year=year, month=month+1, day=0})
+    local month_days   = d.day
+    local column_fday  = (d.wday - d.day + 1 - week_start ) % 7
+
+    --local flags = {"header", "weekdays", "weeknumber", "normal", "focus"}
+    local t, i, j, w, flag, text, markup
+
+    -- Header
+    flag = "header"
+    t = os.time{year=year, month=month, day=1}
+    text = os.date(props.hide_year and "%B" or "%B %Y", t)
+    w = props.fn_embed(make_cell(text, props.font, flag=="yearheader" or flag=="header"), flag)
+    layout:add_widget_at(w, 1, 1, 1, num_columns)
+
+    -- Week days
+    i = 2
+    for j = start_column, num_columns do  -- columns
+        flag = "weekdays"
+        t = os.time{year=2006, month=1, day=j-start_column+week_start}
+        text = string.sub(os.date("%a", t), 1, 2)
+        w = props.fn_embed(make_cell(text, props.font, flag=="yearheader" or flag=="header"), flag)
+        layout:add_widget_at(w, i, j, 1, 1)
+    end
+
+    -- Days
+    i = start_row
+    j = column_fday + start_column
+    local current_week = nil
+    for d=1, month_days do
+        t = os.time{year=year, month=month, day=d}
+        if props.week_numbers then -- Week number
+            text = os.date("%V", t)
+            if tonumber(text) ~= current_week then
+                flag = "weeknumber"
+                w = props.fn_embed(make_cell(text, props.font, flag=="yearheader" or flag=="header"), flag)
+                layout:add_widget_at(w, i, 1, 1, 1)
+                current_week = tonumber(text)
+                if j < start_column then j = start_column end
+            end
+        end
+        -- Normal day
+        flag = "normal"
+        -- Focus day
+        if day and day == d then flag = "focus" end
+        text = string.format("%2d", d)
+        text = flag=="focus" and "<b>"..text.."</b>" or text
+        w = props.fn_embed(make_cell(text, props.font, flag=="yearheader" or flag=="header"), flag)
+        layout:add_widget_at(w, i, j, 1, 1)
+
+        -- find next cell
+        i,j = layout:get_next_empty(i,j)
+    end
+    return props.fn_embed(layout, "month")
+end
+
+
+--- Create a grid layout for the year calendar
+-- @tparam table props Table of year calendar properties
+-- @param date Year to display (number or string)
+-- @treturn widget Grid layout
+local function create_year(props, date)
+    local _, _, year = parse_date(date)
+
+    -- Create a grid widget with the 12 months
+    local in_layout = base.make_widget_declarative{
+        layout      = grid,
+        expand      = true,
+        homogeneous = true,
+        spacing     = 2*props.spacing,
+        forced_num_cols = 4,
+        forced_num_rows = 3,
+    }
+
+    local month_date = ""
```

Comment:
luacheck complaint:

    lib/wibox/widget/calendar.lua:241:11: value assigned to variable month_date is unused